### PR TITLE
elu function implementation

### LINKED ISSUE #17456: elu
implementing the elu activation function

### PR DESCRIPTION
**Pull request by samthakur587**
Close #17456

### CODE REVIEW

**Comment by samthakur587** [comment]
#17558 this the issue for this pull

**Comment by samthakur587** [comment]
it's ready to review the pull

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by samthakur587** [comment]
hey @hirwa-nshuti there is so much merge conflict happening in pull request so can i make a new pull request with a new branch

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @samthakur587 

Would you resolve the merge conflicts for CI run

Thanks

**Comment by samthakur587** [comment]
Hey @hirwa-nshuti 

I have changed only the 6 file for my PR but to solve the merge conflict the 87 files had changed that are not related to my PR and not it not solving from my side can i make a new PR from a New branch to solve all the issue.

Thank you

**Comment by samthakur587** [comment]
finally all previous commit was reset @hirwa-nshuti  now you can go ahead and review my PR 

thanks

File: ivy/functional/ivy/activations.py [hunk 1]
```diff
@@ -609,3 +610,69 @@ def mish(
...
+) -> ivy.Array:
+    """
+    Apply the elu unit function element-wise.
+
+    Parameters
+    ----------
+    x
+        Input array.
+    alpha
+        sclaer multiple of exponent slop of negative input.
```
Comment (hirwa-nshuti) [inline ivy/functional/ivy/activations.py:636]: There is a typo here in scalar and slope would you look into the rest of docstring

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey would you add the issue number of subtask that you're working on

**Comment by samthakur587** [comment]
the issue number of subtask that i am  working on is Extend Ivy Functional API #3856

**Comment by hirwa-nshuti** [comment]
> the issue number of subtask that i am working on is Extend Ivy Functional API #3856

Would you refer to contributing guide and walk through the way to assign yourself a task https://unify.ai/docs/ivy/overview/contributing/open_tasks.html#open-tasks

**Comment by samthakur587** [comment]
sorry for the mistake i have added the PR number in place of issue number in PR description. Now i edited the PR description with issue number. you can check it.
 
thank you

File: ivy/functional/ivy/activations.py (hunk 1 above)
Comment (samthakur587) [reply ivy/functional/ivy/activations.py]: i have corrected this typo in doc string you can check it now.

thanks

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey,

The tests aren't passing in either backends, also you'll have to impelement everything in the experimental API
https://github.com/unifyai/ivy/tree/master/ivy/functional/ivy/experimental

**Comment by samthakur587** [comment]
hey @hirwa-nshuti previously i was implemented the elu activation function outside the ivy  experimental API now i removed all the method from there and implemented with in ivy experimental api you can check now.

thank you

**Comment by samthakur587** [comment]
image

this test of the activation function silu and its not related to my PR. why it is coming i don't know can you please check why it is coming. what changes do i required to solve this.

thank you

**Comment by samthakur587** [comment]
hey @hirwa-nshuti completed all the changes . now its ready to merge

File: ivy_tests/array_api_testing/test_array_api
[the remark this answers is not in the document]
Comment (samthakur587) [reply ivy_tests/array_api_testing/test_array_api:1]: all the changes are completed. now you can review.

File: ivy/functional/backends/tensorflow/experimental/activations.py [hunk 2]
```diff
@@ -66,3 +66,12 @@ def silu(
     if ivy.exists(out):
         return ivy.inplace_update(out, ret).astype(x.dtype)
     return ivy.astype(ret, x.dtype)
+
+
+@with_supported_dtypes({"2.12.0 and below": ("float",)}, backend_version)
+def elu(x: Tensor, /, *, alpha: float = 1.0, out: Optional[Tensor] = None) -> Tensor:
+    alpha = tf.cast(alpha, x.dtype)
+    ret = tf.cast(tf.where(x > 0, x, alpha * tf.expm1(x)), x.dtype)
```
Comment (hirwa-nshuti) [inline ivy/functional/backends/tensorflow/experimental/activations.py:74]: This line is breaking the tests as we call `tf.math.expm1` not `tf.expm1`

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey some minor changes to get the PR merged and also you need to add paddle backend implementation of this function.

Thanks😊

File: ivy/functional/backends/tensorflow/experimental/activations.py (hunk 2 above)
Comment (samthakur587) [reply ivy/functional/backends/tensorflow/experimental/activations.py]: hey @hirwa-nshuti  i have changed the tf.expm1 to tf.math.expm1 and also implemented the elu function for paddle backend. in the paddle backend i called the elu function from paddle.nn.function class so i implemented the elu function in paddle.nn.functional as well with its test function. you can check and if there any mistake please let me know.

thank you

File: ivy/functional/frontends/paddle/nn/functional/activation.py [hunk 3]
```diff
@@ -193,6 +193,20 @@ def relu_(x, name=None):
     return x
 
 
+@with_supported_dtypes({"2.4.2 and below": ("float32", "float64")}, "paddle")
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/activation.py:196]: Here we need to call `ivy.elu` as we are implementing the same function

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey good job so far,
Just run your tests locally and they're failing for the paddle backend, due to assertion error, would you look into this😊
And check the comments below

**Comment by samthakur587** [comment]
Screenshot 2023-06-28 173223

Hey @hirwa-nshuti i am trying to test the paddle frontend test in local i am getting the keyerror  'cpu' i don't where it is store and why this error is coming. can you please help me come out from this problem.

the command i run is : -  
pytest -vv ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py::test_paddle_elu

thank you

File: ivy/functional/frontends/paddle/nn/functional/activation.py [hunk 4]
```diff
@@ -193,6 +193,18 @@ def relu_(x, name=None):
...
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/activation.py:196]: We now use the paddle latest version `2.5.0`

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
With the comment I added will fix the tests for four backends and then you'll need to fix value test error in the paddle backend implementation of the function🙂

File: ivy/functional/frontends/paddle/nn/functional/activation.py (hunk 4 above)
Comment (samthakur587) [reply ivy/functional/frontends/paddle/nn/functional/activation.py]: done , changed the paddle version to 2.5.0

**Comment by samthakur587** [comment]
hey @hirwa-nshuti i have run all the tests locally all are passing but one warring has encounter from numpy backed expm1 function that is show here . 

image

do you has any suggestion to overcome from this. otherwise the PR was completed and ready to merge.

thank you

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
if you run `pytest -v ivy_tests/test_ivy/test_functional/te
st_experimental/test_nn/test_activations.py::test_elu` 
there is failure for paddle backend other changes looks great, just a fix is needed for the paddle backend

**Comment by samthakur587** [comment]
image

Finally all tests are passing locally. Now its Ready to merge . 

thank you

File: ivy/functional/frontends/paddle/nn/functional/activation.py [hunk 5]
```diff
@@ -193,6 +193,18 @@ def relu_(x, name=None):
...
+@with_supported_dtypes({"2.5.0 and below": ("float32", "float64")}, "paddle")
+@to_ivy_arrays_and_back
+def elu(
+    x,
+    /,
+    *,
+    alpha=1.0,
+    name=None,
+):
+    return ivy.elu(x, alpha)
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/activation.py:205]:
```suggestion
    return ivy.elu(x, alpha=alpha)
```

**Review by hirwa-nshuti — APPROVED** [review]
Hey @samthakur587 

All looks good now, and I'm merging this PR😅
Thanks for contributing😊

**Comment by samthakur587** [comment]
Hey @hirwa-nshuti 

That's awesome ! I really appreciate all the help and guidance you've provided throughout this process. It wouldn't have been possible without your valuable contributions.

 Thanks a ton for your support! 😄

**Pull request merged**